modules/internal/DefaultRectangular.chpl

```diff
@@ -705,6 +716,164 @@ module DefaultRectangular {
     }
   }
 
+  inline proc _remoteAccessData.getRADDataIndex(param stridable, ind : idxType) {
+    return this.getRADDataIndex(stridable, chpl__tuplify(ind));
+  }
+
+  inline proc _remoteAccessData.getRADDataIndex(param stridable, ind: rank*idxType) {
+      param chunkify = !defRectSimpleDData;
+
+      if stridable {
+        inline proc chunked_dataIndex(sum, str) {
+          if mdNumChunks == 1 {
+            return (0, sum);
+          } else {
+            const chunk = mdInd2Chunk(ind(mdParDim));
+            return (chunk, sum - mData(chunk).dataOff);
+          }
+        }
+
+        var sum = origin;
+        for param i in 1..rank do
+          sum += (ind(i) - off(i)) * blk(i) / abs(str(i)):idxType;
+        if chunkify then
+          return chunked_dataIndex(sum, str=abs(str(mdParDim)):idxType);
+        else
+          return sum;
+      } else {
+        inline proc chunked_dataIndex(sum) {
+          if mdNumChunks == 1 {
+            return (0, sum);
+          } else {
+            const chunk = mdInd2Chunk(ind(mdParDim));
+            return (chunk, sum - mData(chunk).dataOff);
+          }
+        }
+
+        // optimize common case to get cleaner generated code
+        if (rank == 1 && earlyShiftData) {
+          if blkChanged {
+            if chunkify then
+              return chunked_dataIndex(ind(1) * blk(1));
+            else
+              return ind(1) * blk(1);
+          } else {
+            if chunkify then
+              return chunked_dataIndex(ind(1));
+            else
+              return ind(1);
+          }
+        } else {
+          var sum = if earlyShiftData then 0:idxType else origin;
+
+          if blkChanged {
+            for param i in 1..rank {
+              sum += ind(i) * blk(i);
+            }
+          } else {
+            for param i in 1..rank-1 {
+              sum += ind(i) * blk(i);
+            }
+            sum += ind(rank);
+          }
+
+          if !earlyShiftData then sum -= factoredOffs;
+          if chunkify then
+            return chunked_dataIndex(sum);
+          else
+            return sum;
+        }
+      }
+  }
+
+  proc _remoteAccessData.computeFactoredOffs() {
+    for param i in 1..rank do {
+      factoredOffs = factoredOffs + blk(i) * off(i);
+    }
+  }
+
+  proc _remoteAccessData.initShiftedData() {
+    if earlyShiftData && !stridable {
+      type idxSignedType = chpl__signedType(idxType);
+      const shiftDist = if isIntType(idxType) then origin - factoredOffs
+                        else origin:idxSignedType - factoredOffs:idxSignedType;
+      shiftedData = _ddata_shift(eltType, data, shiftDist);
+    }
+  }
+
+  proc _remoteAccessData.toSlice(newDom) {
+    compilerAssert(defRectSimpleDData && this.rank == newDom.rank);
```

Comment:
Will this assert lead to NUMA failures? Or is this RAD stuff disabled there already?